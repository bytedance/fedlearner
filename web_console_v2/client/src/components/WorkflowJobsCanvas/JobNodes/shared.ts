import { ChartNodeStatus, GlobalNodeRawData, NodeData } from '../types';
import pendingIcon from 'assets/icons/workflow-pending.svg';
import completetdIcon from 'assets/icons/workflow-completed.svg';
import warningIcon from 'assets/icons/workflow-warning.svg';
import errorIcon from 'assets/icons/workflow-error.svg';
import i18n from 'i18n';
import { NodeComponentProps } from 'react-flow-renderer';

export const statusIcons: Record<ChartNodeStatus, string> = {
  [ChartNodeStatus.Pending]: '',
  [ChartNodeStatus.Processing]: pendingIcon,
  [ChartNodeStatus.Success]: completetdIcon,
  [ChartNodeStatus.Warning]: warningIcon,
  [ChartNodeStatus.Error]: errorIcon,
};

export const configStatusText: Record<ChartNodeStatus, string> = {
  [ChartNodeStatus.Pending]: i18n.t('workflow.job_node_pending'),
  [ChartNodeStatus.Processing]: i18n.t('workflow.job_node_configuring'),
  [ChartNodeStatus.Success]: i18n.t('workflow.job_node_config_completed'),
  [ChartNodeStatus.Warning]: i18n.t('workflow.job_node_unfinished'),
  [ChartNodeStatus.Error]: i18n.t('workflow.job_node_invalid'),
};

export const executionStatusText: Record<ChartNodeStatus, string> = {
  [ChartNodeStatus.Pending]: i18n.t('workflow.job_node_waiting'),
  [ChartNodeStatus.Processing]: i18n.t('workflow.job_node_running'),
  [ChartNodeStatus.Success]: i18n.t('workflow.job_node_success'),
  [ChartNodeStatus.Warning]: i18n.t('workflow.job_node_failed'),
  [ChartNodeStatus.Error]: i18n.t('workflow.job_node_stop_running'),
};

export interface JobNodeProps extends NodeComponentProps {
  data: NodeData;
}

export interface GlobalNodeProps extends NodeComponentProps {
  data: NodeData<GlobalNodeRawData>;
}

export const WORKFLOW_JOB_NODE_CHANNELS = {
  change_inheritance: 'job_node.change_inheritance',
  disable_job: 'job_node.disable_job',
  click_add_job: 'job_node.click_add_job',
};
