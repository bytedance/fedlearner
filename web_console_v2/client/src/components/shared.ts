/* istanbul ignore file */
import {
  fetchAlgorithmProjectFileTreeList,
  fetchAlgorithmProjectFileContentDetail,
  fetchAlgorithmFileTreeList,
  fetchAlgorithmFileContentDetail,
  fetchPendingAlgorithmFileTreeList,
  fetchPendingAlgorithmFileContentDetail,
  fetchPeerAlgorithmFileTreeList,
  fetchPeerAlgorithmProjectFileContentDetail,
} from 'services/algorithm';

export function getAlgorithmProjectProps(props: { id: ID }) {
  const { id } = props;

  return {
    id: id,
    isAsyncMode: true,
    getFileTreeList: () => fetchAlgorithmProjectFileTreeList(id!).then((res) => res.data || []),
    getFile: (filePath: string) =>
      fetchAlgorithmProjectFileContentDetail(id!, {
        path: filePath,
      }).then((res) => res.data.content),
  };
}
export function getAlgorithmProps(props: { id: ID }) {
  const { id } = props;

  return {
    id: id,
    isAsyncMode: true,
    getFileTreeList: () => fetchAlgorithmFileTreeList(id!).then((res) => res.data || []),
    getFile: (filePath: string) =>
      fetchAlgorithmFileContentDetail(id!, {
        path: filePath,
      }).then((res) => res.data.content),
  };
}

export function getPendingAlgorithmProps(props: { projId: ID; id: ID }) {
  const { id, projId } = props;
  return {
    id,
    isAsyncMode: true,
    getFileTreeList: () =>
      fetchPendingAlgorithmFileTreeList(projId!, id!).then((res) => res.data || []),
    getFile: (filePath: string) =>
      fetchPendingAlgorithmFileContentDetail(projId!, id!, { path: filePath }).then(
        (res) => res.data.content,
      ),
  };
}

export function getPeerAlgorithmProps(props: { projId: ID; participantId: ID; id: ID; uuid: ID }) {
  const { id, participantId, projId, uuid } = props;
  return {
    id,
    isAsyncMode: true,
    getFileTreeList: () =>
      fetchPeerAlgorithmFileTreeList(projId!, participantId, uuid!).then((res) => res.data || []),
    getFile: (filePath: string) =>
      fetchPeerAlgorithmProjectFileContentDetail(projId!, participantId, uuid!, {
        path: filePath,
      }).then((res) => res.data.content),
  };
}
