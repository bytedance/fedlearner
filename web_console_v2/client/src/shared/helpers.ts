/**
 * @param time time in ms
 */
export function sleep(time: number): Promise<null> {
  return new Promise((resolve) => {
    setTimeout(resolve, time)
  })
}
