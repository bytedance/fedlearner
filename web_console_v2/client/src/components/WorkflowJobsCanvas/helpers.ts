import { XYPosition, Edge } from 'react-flow-renderer';
import { Job, JobState } from 'typings/job';
import { isHead, isLast } from 'shared/array';
import { Dictionary, head, isEmpty, isNil, last } from 'lodash';
import { Variable } from 'typings/variable';
import i18n from 'i18n';
import {
  JobNodeRawData,
  AllNodeTypes,
  ChartNode,
  ChartElements,
  ChartNodeStatus,
  JobNode,
  GlobalConfigNode,
} from './types';

export const NODE_WIDTH = 200;
export const NODE_HEIGHT = 80;
export const GLOBAL_CONFIG_NODE_SIZE = 120;
export const NODE_GAP = 40;

export type ConvertParams = {
  /** Job defintions & Job exe details & some meta-infos such as reused */
  jobs: JobNodeRawData[];
  /** a.k.a. worlflow global settings */
  variables: Variable[];
  /** Extra data pass to react-flow node data */
  data: Dictionary<any>;
};

export type NodeOptions = {
  type: AllNodeTypes;
  selectable: boolean;
};

export type RawDataCol = { raw: JobNodeRawData | Variable[]; isGlobal?: boolean };
export type RawDataRows = Array<RawDataCol[]>;

type ExtraNodeData = {
  index: number;
  status: ChartNodeStatus;
  [key: string]: any;
};

type NodeProcessors = {
  createJob(job: any, extraData: ExtraNodeData, options: NodeOptions): JobNode;
  createGlobal(variables: any, extraData: ExtraNodeData, options: NodeOptions): GlobalConfigNode;
  groupRows(params: ConvertParams): RawDataRows;
};

/**
 * Turn workflow config schema to react-flow elements (will include edges),
 * NOTE: globalVariables is considered as a Node as well
 */
export function convertToChartElements(
  params: ConvertParams,
  options: NodeOptions,
  processors: NodeProcessors = {
    createGlobal: _createGlobalNode,
    createJob: _createJobNode,
    groupRows: groupByDependencies,
  },
): ChartElements {
  /**
   * 1. Group jobs/global-vars into rows
   */
  const rows = processors.groupRows(params);

  /**
   * 2. Turn each col-item in rows to chart node
   */
  let selfIncreaseIndex = 0;
  const nodesRows = rows.map((row) => {
    return row.map((col) => {
      const extraData = { index: selfIncreaseIndex++, rows } as any;

      // Process params.data
      Object.entries(params.data || {}).forEach(([key, value]) => {
        if (typeof value === 'function') {
          return (extraData[key] = value(col));
        }

        extraData[key] = value;
      });

      if (col.isGlobal) {
        return processors.createGlobal(col.raw as Variable[], extraData, options);
      }
      return processors.createJob(col.raw as JobNodeRawData, extraData, options);
    });
  });

  /**
   * 3. Calculate Node position & Generate Edges
   */
  const jobsCountInBiggestRow = Math.max(...nodesRows.map((r) => r.length));
  const maxWidth = _getRowWidth(jobsCountInBiggestRow);
  const midlineX = maxWidth / 2;

  return nodesRows.reduce((result, curRow, rowIdx) => {
    const isHeadRow = isHead(curRow, nodesRows);

    curRow.forEach((node, nodeIdx) => {
      const position = _getNodePosition({
        nodesCount: curRow.length,
        rowIdx,
        nodeIdx,
        midlineX,
        isGlobal: node.data.isGlobal,
      });

      Object.assign(node.position, position);

      if (!isHeadRow) {
        // Create Edges
        // NOTE: only rows not at head can have edge
        const prevRow = nodesRows[rowIdx - 1];

        if (isHead(node, curRow)) {
          const source = head(prevRow)!;
          source.data.isSource = true;
          node.data.isTarget = true;
          result.push(_createEdge(source, node));
        }

        if (isLast(node, curRow) && (prevRow.length > 1 || curRow.length > 1)) {
          const source = last(prevRow)!;
          source.data.isSource = true;
          node.data.isTarget = true;
          result.push(_createEdge(source, node));
        }
      }

      result.push(node);
    });
    return result;
  }, [] as ChartElements);
}

export function groupByDependencies(params: ConvertParams) {
  const { jobs, variables } = params;
  const hasGlobalVars = !isNil(variables) && !isEmpty(variables);

  // 1. Group Jobs to rows by dependiences
  // e.g. Say we have jobs like [1, 2, 3, 4, 5] in which 2,3,4 is depend on 1, 5 depend on 2,3,4
  // we need to render jobs like charts below,
  //
  // row-1        █1█
  //          ╭┈┈┈ ↓ ┈┈┈╮
  // row-2   █2█  █3█  █4█
  //          ╰┈┈┈ ↓ ┈┈┈╯
  // row-3        █5█
  //
  // thus the processed rows will looks like [[1], [2, 3, 4], [5]]
  const rows: RawDataRows = [];
  let rowIdx = 0;

  // If global variables existing, always put it into first row
  if (hasGlobalVars) {
    rows.push([{ raw: variables!, isGlobal: true }]);
    rowIdx++;
  }

  return jobs.reduce((rows, job) => {
    if (shouldPutIntoNextRow()) {
      rowIdx++;
    }

    addANewRowIfNotExist();

    rows[rowIdx].push({ raw: job });

    return rows;

    /**
     * When the Job depend on some Jobs before it (Pre-jobs),
     * plus the Pre-job(s) ARE on the current row,
     * then we should put it into next row
     */
    function shouldPutIntoNextRow() {
      if (!job.dependencies) return false;

      try {
        return job.dependencies.some((dep) => {
          return rows[rowIdx].some((item) => {
            const raw = item.raw as JobNodeRawData;

            return dep.source === raw.name;
          });
        });
      } catch (error) {
        return false;
      }
    }
    function addANewRowIfNotExist() {
      if (!rows[rowIdx]) {
        rows[rowIdx] = [];
      }
    }
  }, rows);
}

/* istanbul ignore next */
export function convertExecutionStateToStatus(state: JobState): ChartNodeStatus {
  return {
    [JobState.NEW]: ChartNodeStatus.Pending,
    [JobState.WAITING]: ChartNodeStatus.Pending,
    [JobState.RUNNING]: ChartNodeStatus.Processing,
    [JobState.COMPLETED]: ChartNodeStatus.Success,
    [JobState.STOPPED]: ChartNodeStatus.Error,
    [JobState.FAILED]: ChartNodeStatus.Warning,
    [JobState.INVALID]: ChartNodeStatus.Warning,
  }[state];
}

// We using the name of job as Node ID at present
export function getNodeIdByJob(job: Job) {
  return job.name;
}

// --------------- Private helpers  ---------------

function _createGlobalNode(
  variables: Variable[],
  extraData: ExtraNodeData,
  options: NodeOptions,
): GlobalConfigNode {
  const name = i18n.t('workflow.label_global_config');
  const { type, ...restOptions } = options;

  return {
    id: name,
    type: 'global',
    ...restOptions,
    data: {
      raw: {
        variables,
        name,
      },
      isGlobal: true,
      ...extraData,
    },
    position: { x: 0, y: 0 },
  };
}

function _createJobNode(
  job: JobNodeRawData,
  extraData: ExtraNodeData,
  options: NodeOptions,
): JobNode {
  const isFork = options?.type === 'fork';

  return {
    id: getNodeIdByJob(job),
    ...options,
    data: {
      raw: job,
      ...extraData,
      mark: job.mark || undefined,
      inherited: isFork ? false : undefined, // under fork mode defaults to false
    },
    position: { x: 0, y: 0 }, // position will be calculated in later step
  };
}

function _createEdge(source: ChartNode, target: ChartNode): Edge {
  return {
    id: `E|${source.id}-${target.id}`,
    source: source.id,
    target: target.id,
    type: 'smoothstep',
  };
}

function _getRowWidth(count: number) {
  return count * NODE_WIDTH + (count - 1) * NODE_GAP;
}

type GetPositionParams = {
  nodesCount: number;
  rowIdx: number;
  nodeIdx: number;
  midlineX: number;
  isGlobal?: boolean;
};

function _getNodePosition({
  nodesCount,
  rowIdx,
  nodeIdx,
  midlineX,
  isGlobal,
}: GetPositionParams): XYPosition {
  const _1stNodeX = midlineX - (NODE_WIDTH * nodesCount) / 2 - ((nodesCount - 1) * NODE_GAP) / 2;
  let x = _1stNodeX + nodeIdx * (NODE_WIDTH + NODE_GAP);

  const y = isGlobal
    ? (NODE_HEIGHT - GLOBAL_CONFIG_NODE_SIZE) / 1.5
    : rowIdx * (NODE_HEIGHT + NODE_GAP);

  if (isGlobal) {
    x += (NODE_WIDTH - GLOBAL_CONFIG_NODE_SIZE) / 2;
  }

  return { x, y };
}
