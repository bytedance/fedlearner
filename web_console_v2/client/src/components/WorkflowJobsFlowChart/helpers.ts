import { XYPosition, Edge } from 'react-flow-renderer';
import { Job, JobState } from 'typings/job';
import { isHead, isLast } from 'shared/array';
import { head, isEmpty, isNil, last } from 'lodash';
import { Variable } from 'typings/variable';
import i18n from 'i18n';
import {
  NodeDataRaw,
  ChartNodeType,
  ChartNode,
  ChartElements,
  JobNodeStatus,
  JobNode,
  GlobalConfigNode,
} from './types';

const TOP_OFFSET = 100;
const LEFT_OFFSET = 100;

export const NODE_WIDTH = 200;
export const NODE_HEIGHT = 80;
export const GLOBAL_CONFIG_NODE_SIZE = 120;
export const NODE_GAP = 30;

/**
 * Turn job defintitions to flow elements (include edges),
 * NOTE: globalVariables is considered as a Node as well
 */
export function convertToChartElements(
  {
    jobs,
    globalVariables,
    side,
  }: { jobs: NodeDataRaw[]; globalVariables?: Variable[]; side?: string },
  options: { type: ChartNodeType; selectable: boolean },
): ChartElements {
  const hasGlobalVars = !isNil(globalVariables) && !isEmpty(globalVariables);

  // 1. Group Jobs to rows by dependiences
  // e.g. Say we have jobs like [1, 2, 3, 4, 5] in which 2,3,4 is depend on 1, 5 depend on 2,3,4
  // we need to render jobs like charts below,
  //
  // row-1        █1█
  //          ╭┈┈┈ ↓ ┈┈┈╮
  // row-2   █2█  █3█  █4█
  //          ╰┈┈┈ ↓ ┈┈┈╯
  // row-3        █5█
  //
  // thus the processed rows will looks like [[1], [2, 3, 4], [5]]
  const rows: Array<ChartNode[]> = [];
  let rowIdx = 0;

  // If global variables existing, always put it into first row
  if (hasGlobalVars) {
    const globalNode = _createGlobalConfigNode({ variables: globalVariables!, options, side });
    rows.push([globalNode]);
    rowIdx++;
  }

  const rowsComputed = jobs.reduce((rows, job, jobIdx) => {
    if (shouldPutIntoNextRow()) {
      rowIdx++;
    }
    addANewRowIfNotExist();

    const node = _createJobNode({ job, index: jobIdx, options, hasGlobalVars, side });

    pushToCurrentRow(node);

    return rows;

    /**
     * When the Job depend on some Jobs before it (Pre-jobs),
     * plus the Pre-job(s) ARE on the current row,
     * then we should put it into next row
     */
    function shouldPutIntoNextRow() {
      if (!job.dependencies) return false;

      return job.dependencies.some((dep) => {
        return rows[rowIdx].some((item) => item.id === dep.source);
      });
    }
    function pushToCurrentRow(node: JobNode) {
      rows[rowIdx].push(node);
    }
    function addANewRowIfNotExist() {
      if (!rows[rowIdx]) {
        rows[rowIdx] = [];
      }
    }
  }, rows);

  // 2. Calculate Node position & Generate Edges
  const jobsCountInMostBigRow = Math.max(...rowsComputed.map((r) => r.length));
  const maxWidth = _getRowWidth(jobsCountInMostBigRow);
  const midlineX = maxWidth / 2 + LEFT_OFFSET;

  return rowsComputed.reduce((result, curRow, rowIdx) => {
    const isHeadRow = isHead(curRow, rowsComputed);

    curRow.forEach((node, nodeIdx) => {
      const position = _getNodePosition({
        nodesCount: curRow.length,
        rowIdx,
        nodeIdx,
        midlineX,
        type: node.type,
      });

      Object.assign(node.position, position);

      if (!isHeadRow) {
        // Create Edges
        // NOTE: only rows not at head can have edge
        const prevRow = rowsComputed[rowIdx - 1];

        if (isHead(node, curRow)) {
          const source = head(prevRow)!;
          source.data.isSource = true;
          node.data.isTarget = true;
          result.push(_createEdge(source, node));
        }

        if (isLast(node, curRow) && (prevRow.length > 1 || curRow.length > 1)) {
          const source = last(prevRow)!;
          source.data.isSource = true;
          node.data.isTarget = true;
          result.push(_createEdge(source, node));
        }
      }

      result.push(node);
    });
    return result;
  }, [] as ChartElements);
}

export function convertExecutionStateToStatus(state: JobState): JobNodeStatus {
  return {
    [JobState.NEW]: JobNodeStatus.Pending,
    [JobState.WAITING]: JobNodeStatus.Pending,
    [JobState.RUNNING]: JobNodeStatus.Processing,
    [JobState.COMPLETE]: JobNodeStatus.Success,
    [JobState.STOPPED]: JobNodeStatus.Error,
    [JobState.FAILED]: JobNodeStatus.Warning,
    [JobState.INVALID]: JobNodeStatus.Warning,
  }[state];
}

// We using the name of job as Node ID at present
export function getNodeIdByJob(job: Job) {
  return job.name;
}

// --------------- Private helpers  ---------------

function _createGlobalConfigNode({
  variables,
  options,
  side,
}: {
  variables: Variable[];
  options?: any;
  side?: string;
}): GlobalConfigNode {
  const name = i18n.t('workflow.label_global_config');
  const isFork = options?.type === 'fork';

  return {
    id: name,
    type: 'global',
    data: {
      raw: ({
        variables,
        name,
        dependencies: [],
      } as unknown) as NodeDataRaw,
      index: 0,
      // When fork type, inherit initially set to true, status to Success
      status: isFork ? JobNodeStatus.Success : JobNodeStatus.Pending,
      inherit: isFork,
      side,
    },
    position: { x: 0, y: 0 },
  };
}

function _createJobNode(params: {
  job: NodeDataRaw;
  index: number;
  options: any;
  hasGlobalVars?: boolean;
  side?: string;
}): JobNode {
  const { job, index, options, hasGlobalVars, side } = params;
  const isFork = options?.type === 'fork';
  const status = job.state
    ? convertExecutionStateToStatus(job.state)
    : isFork
    ? JobNodeStatus.Success
    : JobNodeStatus.Pending;

  return {
    id: getNodeIdByJob(job),
    data: {
      raw: job,
      index: hasGlobalVars ? index + 1 : index, // if have global variables, all nodes should put after it
      mark: job.mark || undefined,
      status,
      inherit: isFork,
      side,
    },
    position: { x: 0, y: 0 }, // position will be calculated in later step
    ...options,
  };
}

function _createEdge(source: ChartNode, target: ChartNode): Edge {
  return {
    id: `E|${source.id}-${target.id}`,
    source: source.id,
    target: target.id,
    type: 'smoothstep',
  };
}

function _getRowWidth(count: number) {
  return count * NODE_WIDTH + (count - 1) * NODE_GAP;
}

type GetPositionParams = {
  nodesCount: number;
  rowIdx: number;
  nodeIdx: number;
  midlineX: number;
  type: ChartNodeType;
};
function _getNodePosition({
  nodesCount,
  rowIdx,
  nodeIdx,
  midlineX,
  type,
}: GetPositionParams): XYPosition {
  const isGlobalNode = type === 'global';
  const _1stNodeX = midlineX - (NODE_WIDTH * nodesCount) / 2 - ((nodesCount - 1) * NODE_GAP) / 2;
  let x = _1stNodeX + nodeIdx * (NODE_WIDTH + NODE_GAP);

  const y =
    TOP_OFFSET +
    (isGlobalNode
      ? (NODE_HEIGHT - GLOBAL_CONFIG_NODE_SIZE) / 1.5
      : rowIdx * (NODE_HEIGHT + NODE_GAP));

  if (type === 'global') {
    x += (NODE_WIDTH - GLOBAL_CONFIG_NODE_SIZE) / 2;
  }

  return { x, y };
}
