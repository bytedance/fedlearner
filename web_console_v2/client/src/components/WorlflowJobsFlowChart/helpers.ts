import { Elements, Node, XYPosition, Edge } from 'react-flow-renderer';
import { Job } from 'typings/workflow';
import { isHead, isLast } from 'shared/array';
import { head, last } from 'lodash';

const TOP_OFFSET = 100;
const LEFT_OFFSET = 30;
export const NODE_WIDTH = 200;
export const NODE_HEIGHT = 80;
export const NODE_GAP = 30;

export enum JobNodeStatus {
  Pending,
  Configuring,
  Unfinished,
  Completed,
}

export type JobNodeData = {
  raw: Job; // each jobs raw-config
  index: number;
  isSource?: boolean;
  isTarget?: boolean;
  status: JobNodeStatus;
  values?: object;
};
export interface JobNode extends Node {
  data: JobNodeData;
}

export function convertJobsToElements(jobs: Job[]): any {
  // 1. Group Jobs to rows by dependiences
  // e.g. Say we have jobs like [1, 2, 3, 4, 5] in which 2,3,4 is depend on 1, 5 depend on 2,3,4
  // we need to render jobs like charts below,
  //
  // row-1        █1█
  //          ╭┈┈┈ ↓ ┈┈┈╮
  // row-2   █2█  █3█  █4█
  //          ╰┈┈┈ ↓ ┈┈┈╯
  // row-3        █5█
  //
  // thus the processed rows will looks like [[1], [2, 3, 4], [5]]
  let rowIdx = 0;
  const rowsComputed = jobs.reduce((rows: Array<JobNode[]>, job, jobIdx) => {
    if (shouldPutIntoNextRow()) {
      rowIdx++;
    }
    addANewRowIfNotExist();

    const node = createNode();

    pushToCurrentRow(node);

    return rows;

    /**
     * When the Job depend on some Jobs before it (Pre-jobs),
     * plus the Pre-job(s) ARE on the current row,
     * then we should put it into next row
     */
    function shouldPutIntoNextRow() {
      if (!job.dependencies) return false;

      return job.dependencies.some((dep) => {
        return rows[rowIdx].some((item) => item.id === dep.source);
      });
    }
    function pushToCurrentRow(node: JobNode) {
      rows[rowIdx].push(node);
    }
    function addANewRowIfNotExist() {
      if (!rows[rowIdx]) {
        rows[rowIdx] = [];
      }
    }

    function createNode(): JobNode {
      return {
        id: getNodeIdByJob(job),
        data: { raw: job, index: jobIdx, status: JobNodeStatus.Pending },
        position: { x: 0, y: 0 }, // position will be calculated in later step
      };
    }
  }, []);

  // 2. Calculate Node position & Generate Edges
  const jobsCountInMostBigRow = Math.max(...rowsComputed.map((r) => r.length));
  const maxWidth = getRowWidth(jobsCountInMostBigRow);
  const midlineX = maxWidth / 2 + LEFT_OFFSET;

  return rowsComputed.reduce((result, curRow, rowIdx) => {
    const isHeadRow = isHead(curRow, rowsComputed);
    const isLastRow = isLast(curRow, rowsComputed);

    curRow.forEach((node, nodeIdx) => {
      const position = getNodePosition({ nodesCount: curRow.length, rowIdx, nodeIdx, midlineX });

      Object.assign(node.position, position);

      if (!isHeadRow) {
        // Create Edges
        // NOTE: only rows not at head can have edge
        const prevRow = rowsComputed[rowIdx - 1];

        if (isHead(node, curRow)) {
          const source = head(prevRow)!;
          source.data.isSource = true;
          node.data.isTarget = true;
          result.push(createEdge(source, node));
        }

        if (isLast(node, curRow) && (prevRow.length > 1 || curRow.length > 1)) {
          const source = last(prevRow)!;
          source.data.isSource = true;
          node.data.isTarget = true;
          result.push(createEdge(source, node));
        }
      } else {
      }

      if (isLastRow) {
        // node.data.source
      }

      result.push(node);
    });
    return result;
  }, [] as Elements);
}

function getRowWidth(count: number) {
  return count * NODE_WIDTH + (count - 1) * NODE_GAP;
}

type GetPositionParams = { nodesCount: number; rowIdx: number; nodeIdx: number; midlineX: number };

function getNodePosition({ nodesCount, rowIdx, nodeIdx, midlineX }: GetPositionParams): XYPosition {
  const _1stNodeX = midlineX - (NODE_WIDTH * nodesCount) / 2 - ((nodesCount - 1) * NODE_GAP) / 2;
  const x = _1stNodeX + nodeIdx * (NODE_WIDTH + NODE_GAP);
  const y = TOP_OFFSET + rowIdx * (NODE_HEIGHT + NODE_GAP);

  return { x, y };
}

// We using the name of job as Node ID at present
export function getNodeIdByJob(job: Job) {
  return job.name;
}

function createEdge(source: JobNode, target: JobNode): Edge {
  return {
    id: `E|${source.id}-${target.id}`,
    source: source.id,
    target: target.id,
    type: 'smoothstep',
  };
}
