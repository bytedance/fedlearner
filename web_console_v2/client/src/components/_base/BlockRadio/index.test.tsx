import React from 'react';
import { render, fireEvent, screen, waitFor } from '@testing-library/react';
import { sleep, getRandomInt } from 'shared/helpers';
import Component from './index';

describe('<BlockRadio />', () => {
  test('default props layout', () => {
    const options = Array(3)
      .fill(0)
      .map((_, index) => {
        return {
          value: index,
          label: `option ${index}`,
        };
      });
    const props = {
      options,
      isCenter: true,
      gap: 77,
    };
    const { container } = render(<Component {...props} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('beforeChange and onChange should be called', async () => {
    const beforeChange = jest.fn();
    const onChange = jest.fn();
    const optionsLength = getRandomInt(10, 20);

    const props = {
      options: generateRandomOptions(optionsLength),
      beforeChange,
      onChange,
    };

    render(<Component {...props} />);
    const blockList = screen.getAllByRole('radio');
    const randomNumGenerator = getRandomIntWrapper(0, optionsLength - 1);
    let clickIndex: number;

    // --------------
    beforeChange.mockReturnValue(Promise.resolve(false));
    clickIndex = randomNumGenerator.next();
    fireEvent.click(blockList[clickIndex]);
    expect(beforeChange).toHaveBeenCalledWith(clickIndex);
    expect(beforeChange).toHaveBeenCalledTimes(1);
    await waitFor(() => {
      expect(onChange).toHaveBeenCalledTimes(0);
    });

    // --------------
    beforeChange.mockReturnValue(Promise.resolve(true));
    clickIndex = randomNumGenerator.next();

    fireEvent.click(blockList[clickIndex]);
    // beforeChange 之前已经执行过一次，所以是 2
    expect(beforeChange).toHaveBeenCalledWith(clickIndex);
    expect(beforeChange).toHaveBeenCalledTimes(2);
    await waitFor(() => {
      expect(onChange).toHaveBeenCalledWith(clickIndex);
      expect(onChange).toHaveBeenCalledTimes(1);
    });

    // --------------
    beforeChange.mockReturnValue(sleep(1000).then(() => true));
    clickIndex = randomNumGenerator.next();
    fireEvent.click(blockList[clickIndex]);
    expect(beforeChange).toHaveBeenCalledWith(clickIndex);
    expect(beforeChange).toHaveBeenCalledTimes(3);
    await waitFor(
      () => {
        expect(onChange).toHaveBeenCalledWith(clickIndex);
        expect(onChange).toHaveBeenCalledTimes(2);
      },
      { timeout: 2000 },
    );
  });

  test('renderBlockInner should work', async () => {
    const optionsLength = getRandomInt(10, 20);
    const renderBlockInner = jest.fn((props, options) => {
      return <h1 className={options.isActive ? 'active' : ''}>{props.value}</h1>;
    });
    const props = {
      options: generateRandomOptions(optionsLength),
      renderBlockInner,
    };

    const { rerender } = render(<Component {...props} />);
    expect(renderBlockInner).toHaveBeenCalledTimes(optionsLength);
    expect(renderBlockInner).toHaveBeenCalledWith(
      expect.objectContaining({
        label: expect.any(String),
        value: expect.any(Number),
      }),
      expect.objectContaining({
        label: expect.anything(),
        isActive: expect.any(Boolean),
      }),
    );

    // 如果 value 改变，应该在相应下标传入 isActive
    const selectedIndex = getRandomInt(0, optionsLength - 1);
    rerender(<Component {...props} value={selectedIndex} />);
    const blockList = screen.getAllByRole('radio');
    const activeBlock = blockList[selectedIndex];
    const innerContent = activeBlock.querySelector('.active');
    await waitFor(() => {
      expect(innerContent).toBeTruthy();
    });
  });

  test('set option to disabled should work', async () => {
    const optionsLength = getRandomInt(10, 20);
    const disabledIndex = getRandomInt(0, optionsLength - 1);
    const options = generateRandomOptions(optionsLength);
    const onChange = jest.fn();

    options[disabledIndex].disabled = true;
    const props = {
      options,
      onChange,
    };

    render(<Component {...props} />);

    const blockList = screen.getAllByRole('radio');
    const disabledBlock = blockList[disabledIndex];
    fireEvent.click(disabledBlock);

    await waitFor(() => {
      expect(onChange).toHaveBeenCalledTimes(0);
    });
  });
});

function generateRandomOptions(length = 4) {
  return Array(length)
    .fill(0)
    .map((_, index) => {
      return {
        value: index,
        label: `option ${index}`,
        disabled: false,
      };
    });
}

function getRandomIntWrapper(min = 1, max = 10) {
  const record: number[] = [];

  return {
    next() {
      if (record.length >= max - min) {
        throw new Error('record overflow');
      }

      let tmp = getRandomInt(min, max);
      while (record.includes(tmp) === true) {
        tmp = getRandomInt(min, max);
      }
      record.push(tmp);
      return tmp;
    },
  };
}
