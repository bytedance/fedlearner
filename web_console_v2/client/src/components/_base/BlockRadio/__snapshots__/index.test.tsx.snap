// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BlockRadio /> default props layout 1`] = `
.c2 {
  display: grid;
  grid-gap: 12px;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  grid-auto-columns: auto;
  grid-template-rows: auto;
  grid-auto-flow: column;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  margin-right: -77px;
  margin-bottom: -77px;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c1 {
  --border-color: transparent;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-box-flex: var(--flex-grow,1);
  -webkit-flex-grow: var(--flex-grow,1);
  -ms-flex-positive: var(--flex-grow,1);
  flex-grow: var(--flex-grow,1);
  min-height: 32px;
  padding: 0 12px;
  width: auto;
  margin-right: 77px;
  margin-bottom: 77px;
  border: 1.5px solid var(--border-color,var(--lineColor));
  border-radius: 4px;
  cursor: pointer;
  background-color: var(--bg-color,var(--componentBackgroundColorGray));
}

.c1:hover {
  --label-color: var(--primaryColor);
}

.c1[data-is-active='true'] {
  --label-color: var(--primaryColor);
  --border-color: var(--primaryColor);
  --label-weight: 500;
  --bg-color: #fff;
}

.c1[data-is-disabled='true'] {
  cursor: not-allowed;
  --label-color: var(--textColorDisabled);
}

.c1[data-is-active='true'][data-is-disabled='true'] {
  --label-color: initial;
  --border-color: initial;
  --label-weight: initial;
  --bg-color: #fff;
}

.c3 {
  font-size: 12px;
  color: var(--label-color,var(--textColorStrong));
  font-weight: var(--label-weight,normal);
  cursor: inherit;
}

<div
  class="c0"
>
  <div
    class="c1"
    data-is-active="false"
    role="radio"
    style="--flex-grow: 1;"
  >
    <div
      class="c2"
      role="grid"
    >
      <label
        class="c3"
      >
        option 0
      </label>
    </div>
  </div>
  <div
    class="c1"
    data-is-active="false"
    role="radio"
    style="--flex-grow: 1;"
  >
    <div
      class="c2"
      role="grid"
    >
      <label
        class="c3"
      >
        option 1
      </label>
    </div>
  </div>
  <div
    class="c1"
    data-is-active="false"
    role="radio"
    style="--flex-grow: 1;"
  >
    <div
      class="c2"
      role="grid"
    >
      <label
        class="c3"
      >
        option 2
      </label>
    </div>
  </div>
</div>
`;
